File: api-app/src/main/java/de/muspellheim/activitysampling/api/application/TimesheetProjection.java
// Copyright (c) 2025 Falko Schumann. All rights reserved. MIT license.

package de.muspellheim.activitysampling.api.application;

import de.muspellheim.activitysampling.api.common.Lists;
import de.muspellheim.activitysampling.api.domain.activities.Activity;
import de.muspellheim.activitysampling.api.domain.activities.Calendar;
import de.muspellheim.activitysampling.api.domain.activities.Holiday;
import de.muspellheim.activitysampling.api.domain.activities.TimesheetEntry;
import de.muspellheim.activitysampling.api.domain.activities.TimesheetQuery;
import de.muspellheim.activitysampling.api.domain.activities.TimesheetQueryResult;
import de.muspellheim.activitysampling.api.domain.activities.WorkingHoursSummary;
import de.muspellheim.activitysampling.api.infrastructure.ActivityLoggedEvent;
import java.time.Clock;
import java.time.Duration;
import java.time.Instant;
import java.time.LocalDate;
import java.time.ZoneId;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

class TimesheetProjection {

  private final LocalDate startInclusive;
  private final LocalDate endExclusive;
  private final ZoneId timeZone;
  private final Duration defaultCapacity;
  private final LocalDate today;
  private final Calendar calendar;

  private List<TimesheetEntry> entries = new ArrayList<>();
  private Duration totalHours = Duration.ZERO;
  private Duration capacity;
  private Duration offset;

  TimesheetProjection(
      TimesheetQuery query,
      ActivitiesConfiguration configuration,
      List<Holiday> holidays,
      Clock clock) {
    startInclusive = query.from();
    endExclusive = query.to().plusDays(1);
    timeZone = query.timeZone() != null ? query.timeZone() : ZoneId.systemDefault();
    defaultCapacity = configuration.capacity();
    today = clock.instant().atZone(timeZone).toLocalDate();
    var holidayDates = holidays.stream().map(Holiday::date).collect(Collectors.toSet());
    calendar = new Calendar();
    calendar.initHolidays(holidayDates);
  }

  Instant getStartInclusive() {
    return startInclusive.atStartOfDay(timeZone).toInstant();
  }

  public Instant getEndExclusive() {
    return endExclusive.atStartOfDay(timeZone).toInstant();
  }

  TimesheetQueryResult project(Stream<ActivityLoggedEvent> events) {
    events
        .sorted(Comparator.comparing(ActivityLoggedEvent::timestamp))
        .map(it -> ActivityMapping.map(it, timeZone))
        .forEach(
            it -> {
              updateEntries(it);
              updateTotalHours(it);
            });
    determineCapacity();
    determineOffset();
    sortEntries();
    return TimesheetQueryResult.builder()
        .entries(entries)
        .workingHoursSummary(
            WorkingHoursSummary.builder()
                .totalHours(totalHours)
                .capacity(capacity)
                .offset(offset)
                .build())
        .build();
  }

  private void updateEntries(Activity activity) {
    var activityDate = activity.dateTime().toLocalDate();
    var index =
        Lists.indexOf(
            entries,
            e ->
                e.date().equals(activityDate)
                    && e.client().equals(activity.client())
                    && e.project().equals(activity.project())
                    && e.task().equals(activity.task()));
    if (index == -1) {
      var newEntry =
          TimesheetEntry.builder()
              .date(activity.dateTime().toLocalDate())
              .client(activity.client())
              .project(activity.project())
              .task(activity.task())
              .hours(activity.duration())
              .build();
      entries.add(newEntry);
    } else {
      var existingEntry = entries.get(index);
      var accumulatedHours = existingEntry.hours().plus(activity.duration());
      var updatedEntry = existingEntry.withHours(accumulatedHours);
      entries.set(index, updatedEntry);
    }
  }

  private void updateTotalHours(Activity activity) {
    totalHours = totalHours.plus(activity.duration());
  }

  private void determineCapacity() {
    var businessDays = calendar.countBusinessDays(startInclusive, endExclusive);
    capacity = Duration.ofHours(businessDays * defaultCapacity.toHours() / 5);
  }

  private void determineOffset() {
    LocalDate end;
    if (today.isBefore(startInclusive)) {
      end = startInclusive;
    } else if (today.plusDays(1).isAfter(endExclusive)) {
      end = endExclusive;
    } else {
      end = today.plusDays(1);
    }
    var businessDays = calendar.countBusinessDays(startInclusive, end);
    offset = totalHours.minusHours(businessDays * 8);
  }

  private void sortEntries() {
    entries =
        entries.stream()
            .sorted(
                Comparator.comparing(TimesheetEntry::date)
                    .thenComparing(TimesheetEntry::client)
                    .thenComparing(TimesheetEntry::project)
                    .thenComparing(TimesheetEntry::task))
            .toList();
  }
}

File: src/main/domain/timesheet_projection.ts
// Copyright (c) 2026 Falko Schumann. All rights reserved. MIT license.

import { Temporal } from "@js-temporal/polyfill";

import {
  isTimestampInPeriod,
  normalizeDuration,
} from "../../shared/domain/temporal";
import {
  TimesheetEntry,
  TimesheetQuery,
  TimesheetQueryResult,
} from "../../shared/domain/timesheet_query";
import type { ActivityLoggedEvent } from "./activity_logged_event";
import { Calendar, type Holiday, Vacation } from "./calendar";
import type { Projection } from "./projection";
import { TotalHoursProjection } from "./total_hours_projection";

export class TimesheetProjection implements Projection<TimesheetQueryResult> {
  static create({
    query,
    today = Temporal.Now.plainDateISO("Europe/Berlin"),
    timeZone = "Europe/Berlin",
    capacity = Temporal.Duration.from("PT40H"),
    holidays = [],
    vacations = [],
  }: {
    query: TimesheetQuery;
    today?: Temporal.PlainDate;
    timeZone?: Temporal.TimeZoneLike;
    capacity?: Temporal.Duration;
    holidays?: Holiday[];
    vacations?: Vacation[];
  }) {
    return new TimesheetProjection(
      query,
      today,
      timeZone,
      capacity,
      holidays,
      vacations,
    );
  }

  readonly #query;
  readonly #today;
  readonly #timeZone;
  readonly #capacity;
  readonly #holidays;
  readonly #vacations;
  readonly #timesheetEntryProjection;
  readonly #totalHoursProjection;

  private constructor(
    query: TimesheetQuery,
    today: Temporal.PlainDate,
    timeZone: Temporal.TimeZoneLike,
    capacity: Temporal.Duration,
    holidays: Holiday[],
    vacations: Vacation[],
  ) {
    this.#query = query;
    this.#today = query.today || today;
    this.#timeZone = timeZone;
    this.#capacity = capacity;
    this.#holidays = holidays;
    this.#vacations = vacations;
    this.#timesheetEntryProjection = TimesheetEntryProjection.create({
      timeZone,
    });
    this.#totalHoursProjection = TotalHoursProjection.create();
  }

  update(event: ActivityLoggedEvent) {
    if (
      !isTimestampInPeriod(
        event.timestamp,
        this.#timeZone,
        this.#query.from,
        this.#query.to,
      )
    ) {
      return;
    }

    this.#timesheetEntryProjection.update(event);
    this.#totalHoursProjection.update(event);
  }

  get(): TimesheetQueryResult {
    const entries = this.#timesheetEntryProjection.get();
    const totalHours = this.#totalHoursProjection.get();
    const capacity = this.#determineCapacity(totalHours);
    return TimesheetQueryResult.create({ entries, totalHours, capacity });
  }

  #determineCapacity(totalHours: Temporal.Duration) {
    const calendar = Calendar.create({
      holidays: this.#holidays,
      vacations: this.#vacations,
      capacity: this.#capacity,
    });
    let end: Temporal.PlainDate;
    const { from, to } = this.#query;
    if (Temporal.PlainDate.compare(this.#today, from) < 0) {
      end = from;
    } else if (Temporal.PlainDate.compare(this.#today, to) > 0) {
      end = to;
    } else {
      end = this.#today;
    }
    const businessDays = calendar.countWorkingHours(from, end);
    const offset = totalHours.subtract(businessDays);
    return {
      hours: calendar.countWorkingHours(from, to),
      offset: normalizeDuration(offset),
    };
  }
}

class TimesheetEntryProjection {
  static create({ timeZone }: { timeZone: Temporal.TimeZoneLike }) {
    return new TimesheetEntryProjection(timeZone);
  }

  readonly #timeZone;
  #entries: TimesheetEntry[];

  private constructor(timeZone: Temporal.TimeZoneLike) {
    this.#timeZone = timeZone;
    this.#entries = [];
  }

  update(event: ActivityLoggedEvent) {
    const date = event.timestamp
      .toZonedDateTimeISO(this.#timeZone)
      .toPlainDate();
    const index = this.#entries.findIndex(
      (entry) =>
        Temporal.PlainDate.compare(entry.date, date.toString()) === 0 &&
        entry.client === event.client &&
        entry.project === event.project &&
        entry.task === event.task,
    );
    if (index === -1) {
      const newEntry = TimesheetEntry.create({
        ...event,
        date,
        hours: event.duration,
      });
      this.#entries.push(newEntry);
    } else {
      const existingEntry = this.#entries[index]!;
      const accumulatedHours = existingEntry.hours.add(event.duration);
      this.#entries[index] = TimesheetEntry.create({
        ...existingEntry,
        hours: normalizeDuration(accumulatedHours),
      });
    }
  }

  get() {
    return this.#entries.sort((entry1, entry2) => {
      const dateComparison = Temporal.PlainDate.compare(
        entry1.date,
        entry2.date,
      );
      if (dateComparison !== 0) {
        return dateComparison;
      }
      if (entry1.client !== entry2.client) {
        return entry1.client.localeCompare(entry2.client);
      }
      if (entry1.project !== entry2.project) {
        return entry1.project.localeCompare(entry2.project);
      }
      return entry1.task.localeCompare(entry2.task);
    });
  }
}
